File: dotnet-engine/FormulaEngine/Repositories/PostgresRepository.cs
using System.Text;
using FormulaEngine.Models;
using Npgsql;

namespace FormulaEngine.Repositories;

/// <summary>
/// PostgreSQL implementation of the data repository.
/// Uses Npgsql for database access with optimized batch operations.
/// </summary>
public class PostgresRepository : IDataRepository
{
    private readonly string _connectionString;
    private NpgsqlConnection? _connection;
    private const int BatchSize = 10000;

    public PostgresRepository(string connectionString)
    {
        _connectionString = connectionString;
    }

    /// <summary>
    /// Initialize the repository and establish database connection.
    /// </summary>
    public async Task InitializeAsync()
    {
        _connection = new NpgsqlConnection(_connectionString);
        await _connection.OpenAsync();
        Console.WriteLine("Database connection established.");
    }

    /// <summary>
    /// Get all data records from the t_data table.
    /// Uses streaming for memory efficiency with large datasets.
    /// </summary>
    public async Task<IEnumerable<DataRecord>> GetAllDataRecordsAsync()
    {
        EnsureConnectionOpen();

        var records = new List<DataRecord>();
        const string query = "SELECT data_id, a, b, c, d FROM t_data ORDER BY data_id";

        await using var cmd = new NpgsqlCommand(query, _connection);
        await using var reader = await cmd.ExecuteReaderAsync();

        while (await reader.ReadAsync())
        {
            records.Add(new DataRecord
            {
                DataId = reader.GetInt32(0),
                A = reader.GetDouble(1),
                B = reader.GetDouble(2),
                C = reader.GetDouble(3),
                D = reader.GetDouble(4)
            });
        }

        return records;
    }

    /// <summary>
    /// Get all formulas from the t_targil table.
    /// </summary>
    public async Task<IEnumerable<Formula>> GetAllFormulasAsync()
    {
        EnsureConnectionOpen();

        var formulas = new List<Formula>();
        const string query = "SELECT targil_id, targil, tnai, targil_false FROM t_targil ORDER BY targil_id";

        await using var cmd = new NpgsqlCommand(query, _connection);
        await using var reader = await cmd.ExecuteReaderAsync();

        while (await reader.ReadAsync())
        {
            formulas.Add(new Formula
            {
                TargilId = reader.GetInt32(0),
                Targil = reader.GetString(1),
                Tnai = reader.IsDBNull(2) ? null : reader.GetString(2),
                TargilFalse = reader.IsDBNull(3) ? null : reader.GetString(3)
            });
        }

        return formulas;
    }

    /// <summary>
    /// Save calculation results to the t_results table using bulk insert.
    /// Uses COPY command for optimal performance with large datasets.
    /// </summary>
    public async Task SaveResultsBatchAsync(IEnumerable<CalculationResult> results)
    {
        EnsureConnectionOpen();

        var resultsList = results.ToList();
        if (resultsList.Count == 0) return;

        // Use COPY command for bulk insert (most efficient for PostgreSQL)
        await using var writer = await _connection!.BeginBinaryImportAsync(
            "COPY t_results (data_id, targil_id, method, result) FROM STDIN (FORMAT BINARY)");

        foreach (var result in resultsList)
        {
            await writer.StartRowAsync();
            await writer.WriteAsync(result.DataId, NpgsqlTypes.NpgsqlDbType.Integer);
            await writer.WriteAsync(result.TargilId, NpgsqlTypes.NpgsqlDbType.Integer);
            await writer.WriteAsync(result.Method, NpgsqlTypes.NpgsqlDbType.Varchar);
            
            if (result.Result.HasValue)
            {
                await writer.WriteAsync(result.Result.Value, NpgsqlTypes.NpgsqlDbType.Double);
            }
            else
            {
                await writer.WriteNullAsync();
            }
        }

        await writer.CompleteAsync();
    }

    /// <summary>
    /// Save a log entry with execution timing to the t_log table.
    /// </summary>
    public async Task SaveLogEntryAsync(int targilId, string method, double runTime)
    {
        EnsureConnectionOpen();

        const string query = @"
            INSERT INTO t_log (targil_id, method, run_time)
            VALUES (@targilId, @method, @runTime)";

        await using var cmd = new NpgsqlCommand(query, _connection);
        cmd.Parameters.AddWithValue("targilId", targilId);
        cmd.Parameters.AddWithValue("method", method);
        cmd.Parameters.AddWithValue("runTime", runTime);

        await cmd.ExecuteNonQueryAsync();
    }

    /// <summary>
    /// Clear previous results for a specific calculation method.
    /// </summary>
    public async Task ClearMethodResultsAsync(string method)
    {
        EnsureConnectionOpen();

        const string query = "DELETE FROM t_results WHERE method = @method";

        await using var cmd = new NpgsqlCommand(query, _connection);
        cmd.Parameters.AddWithValue("method", method);

        var deleted = await cmd.ExecuteNonQueryAsync();
        Console.WriteLine($"Cleared {deleted:N0} previous results for method '{method}'.");
    }

    /// <summary>
    /// Clear previous log entries for a specific calculation method.
    /// </summary>
    public async Task ClearMethodLogsAsync(string method)
    {
        EnsureConnectionOpen();

        const string query = "DELETE FROM t_log WHERE method = @method";

        await using var cmd = new NpgsqlCommand(query, _connection);
        cmd.Parameters.AddWithValue("method", method);

        var deleted = await cmd.ExecuteNonQueryAsync();
        Console.WriteLine($"Cleared {deleted:N0} previous log entries for method '{method}'.");
    }

    /// <summary>
    /// Ensure the database connection is open.
    /// </summary>
    private void EnsureConnectionOpen()
    {
        if (_connection == null || _connection.State != System.Data.ConnectionState.Open)
        {
            throw new InvalidOperationException("Database connection is not open. Call InitializeAsync() first.");
        }
    }

    /// <summary>
    /// Dispose of the database connection.
    /// </summary>
    public async ValueTask DisposeAsync()
    {
        if (_connection != null)
        {
            await _connection.CloseAsync();
            await _connection.DisposeAsync();
            _connection = null;
            Console.WriteLine("Database connection closed.");
        }
    }
}

File: python-engine/repositories/postgres_repository.py
"""
PostgreSQL repository for database access.

This module provides async database operations for the formula benchmark
system, including reading data records, formulas, and saving results.
"""

import logging
from typing import List, Optional, AsyncIterator
from contextlib import asynccontextmanager

import asyncpg

import sys
import os
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from models.data_record import DataRecord
from models.formula import Formula
from models.calculation_result import CalculationResult


# Configure module-level logger
logger = logging.getLogger(__name__)


class PostgresRepository:
    """
    Repository for PostgreSQL database operations.
    
    Provides async methods for reading data records and formulas,
    saving calculation results, and logging execution times.
    
    Uses asyncpg for high-performance async PostgreSQL access.
    """
    
    # Default batch size for reading data records (100,000 per batch)
    DEFAULT_BATCH_SIZE = 100_000
    
    def __init__(self, connection_string: str, batch_size: int = DEFAULT_BATCH_SIZE):
        """
        Initialize the repository with connection settings.
        
        Args:
            connection_string: PostgreSQL connection string (asyncpg DSN format)
            batch_size: Number of records to fetch per batch (default: 100,000)
        """
        self._connection_string = connection_string
        self._batch_size = batch_size
        self._pool: Optional[asyncpg.Pool] = None
    
    async def connect(self) -> None:
        """
        Establish connection pool to the database.
        
        Raises:
            ConnectionError: If connection fails
        """
        try:
            self._pool = await asyncpg.create_pool(
                self._connection_string,
                min_size=2,
                max_size=10,
                command_timeout=300  # 5 minute timeout for long operations
            )
            logger.info("Database connection pool established")
        except Exception as e:
            logger.error(f"Failed to connect to database: {e}")
            raise ConnectionError(f"Database connection failed: {e}") from e
    
    async def disconnect(self) -> None:
        """Close the connection pool."""
        if self._pool:
            await self._pool.close()
            self._pool = None
            logger.info("Database connection pool closed")
    
    @asynccontextmanager
    async def get_connection(self) -> AsyncIterator[asyncpg.Connection]:
        """
        Get a connection from the pool as a context manager.
        
        Yields:
            asyncpg.Connection: A database connection
            
        Raises:
            RuntimeError: If pool is not initialized
        """
        if self._pool is None:
            raise RuntimeError("Database pool not initialized. Call connect() first.")
        
        async with self._pool.acquire() as connection:
            yield connection
    
    async def get_all_formulas(self) -> List[Formula]:
        """
        Fetch all formulas from the t_targil table.
        
        Returns:
            List of Formula objects
        """
        query = """
            SELECT targil_id, targil, tnai, targil_false
            FROM t_targil
            ORDER BY targil_id
        """
        
        async with self.get_connection() as conn:
            rows = await conn.fetch(query)
            formulas = [
                Formula(
                    targil_id=row['targil_id'],
                    targil=row['targil'],
                    tnai=row['tnai'],
                    targil_false=row['targil_false']
                )
                for row in rows
            ]
            logger.info(f"Fetched {len(formulas)} formulas from database")
            return formulas
    
    async def get_data_record_count(self) -> int:
        """
        Get the total count of data records.
        
        Returns:
            Total number of records in t_data
        """
        query = "SELECT COUNT(*) FROM t_data"
        
        async with self.get_connection() as conn:
            count = await conn.fetchval(query)
            return count
    
    async def get_data_records_batch(
        self, 
        offset: int, 
        limit: Optional[int] = None
    ) -> List[DataRecord]:
        """
        Fetch a batch of data records from t_data.
        
        Args:
            offset: Starting position (0-indexed)
            limit: Number of records to fetch (default: batch_size)
            
        Returns:
            List of DataRecord objects
        """
        batch_limit = limit or self._batch_size
        
        query = """
            SELECT data_id, a, b, c, d
            FROM t_data
            ORDER BY data_id
            LIMIT $1 OFFSET $2
        """
        
        async with self.get_connection() as conn:
            rows = await conn.fetch(query, batch_limit, offset)
            return [
                DataRecord(
                    data_id=row['data_id'],
                    a=float(row['a']),
                    b=float(row['b']),
                    c=float(row['c']),
                    d=float(row['d'])
                )
                for row in rows
            ]
    
    async def iterate_data_batches(self) -> AsyncIterator[List[DataRecord]]:
        """
        Iterate through all data records in batches.
        
        Yields:
            List[DataRecord]: Batches of data records
            
        This is memory-efficient for processing 1 million records
        by yielding 100,000 records at a time.
        """
        total_count = await self.get_data_record_count()
        offset = 0
        batch_num = 0
        
        logger.info(f"Starting to iterate through {total_count:,} records in batches of {self._batch_size:,}")
        
        while offset < total_count:
            batch = await self.get_data_records_batch(offset)
            if not batch:
                break
            
            batch_num += 1
            logger.debug(f"Fetched batch {batch_num}: {len(batch):,} records (offset: {offset:,})")
            
            yield batch
            offset += len(batch)
        
        logger.info(f"Completed iteration: {batch_num} batches processed")
    
    async def get_all_data_records(self) -> List[DataRecord]:
        """
        Fetch all data records from t_data.
        
        WARNING: This loads all 1 million records into memory.
        Consider using iterate_data_batches() for memory efficiency.
        
        Returns:
            List of all DataRecord objects
        """
        all_records: List[DataRecord] = []
        
        async for batch in self.iterate_data_batches():
            all_records.extend(batch)
        
        logger.info(f"Loaded {len(all_records):,} total data records")
        return all_records
    
    async def clear_method_results(self, method: str) -> int:
        """
        Delete all results for a specific calculation method.
        
        Args:
            method: The method name (e.g., "Python_Eval")
            
        Returns:
            Number of rows deleted
        """
        query = "DELETE FROM t_results WHERE method = $1"
        
        async with self.get_connection() as conn:
            result = await conn.execute(query, method)
            # Extract row count from result string like "DELETE 1000000"
            deleted_count = int(result.split()[-1])
            logger.info(f"Cleared {deleted_count:,} previous results for method '{method}'")
            return deleted_count
    
    async def clear_method_logs(self, method: str) -> int:
        """
        Delete all log entries for a specific calculation method.
        
        Args:
            method: The method name (e.g., "Python_Eval")
            
        Returns:
            Number of rows deleted
        """
        query = "DELETE FROM t_log WHERE method = $1"
        
        async with self.get_connection() as conn:
            result = await conn.execute(query, method)
            deleted_count = int(result.split()[-1])
            logger.info(f"Cleared {deleted_count:,} previous log entries for method '{method}'")
            return deleted_count
    
    async def save_results_batch(self, results: List[CalculationResult]) -> int:
        """
        Bulk insert calculation results to t_results.
        
        Uses PostgreSQL's COPY protocol for efficient bulk insertion.
        
        Args:
            results: List of CalculationResult objects to save
            
        Returns:
            Number of rows inserted
        """
        if not results:
            return 0
        
        # Prepare data as list of tuples
        records = [
            (r.data_id, r.targil_id, r.method, r.result)
            for r in results
        ]
        
        async with self.get_connection() as conn:
            # Use copy_records_to_table for efficient bulk insert
            await conn.copy_records_to_table(
                't_results',
                records=records,
                columns=['data_id', 'targil_id', 'method', 'result']
            )
        
        logger.debug(f"Saved batch of {len(results):,} results")
        return len(results)
    
    async def save_log_entry(
        self, 
        targil_id: int, 
        method: str, 
        run_time: float,
        records_processed: int = 1_000_000
    ) -> int:
        """
        Insert a timing log entry to t_log.
        
        Args:
            targil_id: Formula ID
            method: Calculation method name
            run_time: Execution time in seconds
            records_processed: Number of records processed (default: 1,000,000)
            
        Returns:
            The ID of the inserted log entry
        """
        query = """
            INSERT INTO t_log (targil_id, method, run_time, records_processed)
            VALUES ($1, $2, $3, $4)
            RETURNING log_id
        """
        
        async with self.get_connection() as conn:
            log_id = await conn.fetchval(query, targil_id, method, run_time, records_processed)
            logger.info(
                f"Logged execution: formula {targil_id}, method '{method}', "
                f"time {run_time:.4f}s, records {records_processed:,}"
            )
            return log_id
    
    async def get_results_for_verification(
        self, 
        targil_id: int, 
        limit: int = 100
    ) -> List[dict]:
        """
        Get sample results for a formula to verify across methods.
        
        Args:
            targil_id: Formula ID to check
            limit: Number of data records to compare
            
        Returns:
            List of dicts with results grouped by data_id
        """
        query = """
            SELECT 
                data_id,
                method,
                result
            FROM t_results
            WHERE targil_id = $1
                AND data_id IN (
                    SELECT data_id FROM t_data ORDER BY data_id LIMIT $2
                )
            ORDER BY data_id, method
        """
        
        async with self.get_connection() as conn:
            rows = await conn.fetch(query, targil_id, limit)
            return [dict(row) for row in rows]
